Source/Dafny/Resolver/TypeConstraint.cs

```diff
@@ -0,0 +1,134 @@
+#define TI_DEBUG_PRINT
+//-----------------------------------------------------------------------------
+//
+// Copyright (C) Microsoft Corporation.  All Rights Reserved.
+// Copyright by the contributors to the Dafny Project
+// SPDX-License-Identifier: MIT
+//
+//-----------------------------------------------------------------------------
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Numerics;
+using System.Diagnostics.Contracts;
+using JetBrains.Annotations;
+using Microsoft.BaseTypes;
+using Microsoft.Boogie;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+
+namespace Microsoft.Dafny {
+  public class TypeConstraint {
```

Comment:
Consider changing this class into a record with a primary constructor `record TypeConstraint(Type super, Type sub, ErrorMsg errMsg, bool keepConstraints)`, that's an easy to do refactoring which makes it easy to see from the header of the record what it does.